lib/src/incremental_generator.dart

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/src/generated/engine.dart';
+import 'package:build/build.dart';
+import 'package:path/path.dart' as path;
+
+import 'generator.dart';
+
+/// A generator that will try to generate recursively until 2 consecutive
+/// generations output the same result.
+abstract class IncrementalGenerator extends Generator {
+  final int maxIterations;
+
+  const IncrementalGenerator({this.maxIterations: 1000});
+
+  Future<String> generateForLibraryElement(
+      LibraryElement library, BuildStep buildStep);
+
+  @override
+  Future<String> generate(Element element, BuildStep buildStep) async {
+    if (element is! LibraryElement) return null;
+    final lib = element as LibraryElement;
+
+    // back up the initial content of the part to restore at the end
+    final generatedPart = getGeneratedPart(lib);
+    String genContent, initialContent;
+    try {
+      initialContent = lib.context.getContents(generatedPart.source).data;
+      genContent = initialContent;
+    } on StateError {
+      genContent = '';
+    }
+
+    // generate content several times until 2 consecutive contents are equals
+    int iterationsLeft = maxIterations;
+    while (true) {
+      if (iterationsLeft-- < 0)
+        throw new StateError(
+            'No stable content after $maxIterations generations');
+
+      final nextGenContent = await generateForLibraryElement(lib, buildStep);
+
+      // exit if stable
+      if (nextGenContent == genContent) break;
+
+      genContent = nextGenContent;
+
+      // next increment : add current genContent to initial content
+      lib.context.applyChanges(
```

Comment:
Applying changes to the analyzers view of the world is going to be problematic. As a performance measure we use a single `Resolvers` instance for everything, so this will affect all generators view of the world. This also makes these files available to other builders in the same phase, when it isn't supposed to be available until the next phase. This could lead to some very subtle issues :(.

Also I don't know how this will behave with multiple generators in the same file, I don't think it will end well.